### ISSUE TITLE
Poor performance on mesh with many domains

### ISSUE THREAD

**Issue by florian98765** [firedrakeproject/firedrake#1951]
Dear Firedrake community,

I am currently facing a performance problem when using a mesh `ultra10.msh` which contains many domains (4182).
In our code some problem are only solve on subregions of the full mesh. Thus we often use `dx(index_set)` in order to integrate only over a subdomain. However this seems to lead to very poor performance if the number of domains is very large. See e.g. this example

```
from firedrake import *
import time

mesh = Mesh("ultra10.msh")
V = FunctionSpace(mesh, "CG", 1)
f = Function(V)

t0 = time.time()
x = assemble(f*dx)
t1 = time.time()
print("assemble(f*dx):", t1-t0)

t0 = time.time()
N = tuple(range(1,4183))
x = assemble(f*dx(N))
t1 = time.time()
print("assemble(f*dx(1,2,...,4182)):", t1-t0)
```
[ultra10.msh.gz](https://github.com/firedrakeproject/firedrake/files/5819816/ultra10.msh.gz)

Assembly over the whole mesh takes about 0.06s whereas the assembly with a tuple containing all domains still has not finished after 10 minutes.

Is there a better way of integrating over multiple subdomains?
Would it be possible to add additional tags to e.g. one half of the tuple using dmplex? or does each cell have exactly one id?

best wishes
Florian

**Comment by florian98765** [comment]
Just did some further timings:
```
assemble(f*dx): 0.06059122085571289
assemble(f*dx(1)): 0.19224190711975098
assemble(f*dx(1..10)): 0.8833432197570801
assemble(f*dx(1..100)): 9.025017499923706
assemble(f*dx(1,2,...,4182)): 2586.2544791698456
```
Any ideas?

**Comment by wence-** [comment]
I had a brief profile.

1. There's a quadratic algorithm in the way UFL makes those symbolic forms.

Apply this patch in UFL:
```
diff --git a/ufl/measure.py b/ufl/measure.py
index c624e417..818bcb72 100644
--- a/ufl/measure.py
+++ b/ufl/measure.py
@@ -12,6 +12,8 @@
 
 import numbers
 
+from itertools import chain
+
 from ufl.log import error, deprecate
 from ufl.core.expr import Expr
 from ufl.checks import is_true_ufl_scalar
@@ -411,7 +413,8 @@ class Measure(object):
         # Integral.__add__:
         subdomain_id = self.subdomain_id()
         if isinstance(subdomain_id, tuple):
-            return sum(integrand * self.reconstruct(subdomain_id=d) for d in subdomain_id)
+            return Form(list(chain(*((integrand * self.reconstruct(subdomain_id=d)).integrals()
+                                     for d in subdomain_id))))
 
         # Check that we have an integer subdomain or a string
         # ("everywhere" or "otherwise", any more?)
```

2. The way UFL implements this (and we pick this up) is that it generates basically one form for every subdomain integral: it doesn't try to merge the subdomains based on which integrands are the same. In this particular example, this means that you get N separate integrals, and hence we have to compile N separate kernels (so even if one compile is fast, it's linear in the number of subdomains).

UFL could be smarter here, and merge integrands which are the same up to subdomain_id. In this case, we would go back to having to compile one kernel.

3. The way we execute subdomain_id selected kernels is that for each subdomain, we have a separate assembly loop (over the cells that are taking part). This is again slower than putting everything in one assembly loop and dispatching inside.

In short, there are things we could make faster. One question: in your real application do you have a unique `f` for each subdomain_id?

**Comment by florian98765** [comment]
Dear Lawrence,

I tried to apply your patch directly within our docker container. I simply changed /firedrake/src/ufl/ufl/measure.py.
Unfortunately I cannot confirm any changes in the timings.

Regarding your question: in our application f should be constant for each subdomain. In this particular case we only define different material parameters `mu` on each grains and the solve something like `mu*inner(grad(u), grad(v)*dx(all_grains)`. Sometimes we also have non-magnetic domains which should be excluded from the solve. This is why we done use `everywhere` in general. 

In the meantime I found one workaround which seems to work for our case. Similar to `SubDomainData` we define an `op2.Subset` and then use `dx(mesh, subdomain_data = magnetic_domains)`. Although we still have problems with this approach it seems to fix the timing issue.

Are there any drawbacks when using `subdomain_data`? I think this would be a nice solution, if e.g. there are different subsets for different solvers.

best wishes 
Florian

**Comment by wence-** [comment]
> Are there any drawbacks when using subdomain_data? I think this would be a nice solution, if e.g. there are different subsets for different solvers.

I think that #1956 shows a drawback (subdomain_data doesn't really do what we thought it did).

So in your particular use case, you could probably get good performance by having your grains all marked with the _same_ mesh marker (separate from the non-magnetic domains).

If you need the different markers to be able to specify the different values of mu on the grains, note that the "Cell Set" label inside DMPlex is actually a multi-map, so if you can convince your mesh generator to label grains with two labels (one distinct and one marking "I am a grain"), then we actually pick that up.

Suppose you have three grains with individual labels 1, 2, and 3, as well also giving them the same label 10.

Then

```
expr*dx((1, 2, 3))
# and

expr*dx(10)
```

Will both assemble over all the grains, except that latter will do things in one go, rather than one by one.

**Comment by florian98765** [comment]
Dear Lawrence,
thanks once again for your advice. We now use the multi-map method you proposed.
Everything seems to work and performance is much better.

best wishes
Florian
